// printd/dist/printd.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
	typeof define === 'function' && define.amd ? define(['exports'], factory) :
	(factory((global.printd = {})));
}(this, (function (exports) { 'use strict';

function createStyle(doc, cssText) {
    var style = doc.createElement('style');
    style.type = 'text/css';
    style.appendChild(window.document.createTextNode(cssText));
    return style;
}
function createIFrame(parent) {
    if (parent === void 0) { parent = window.document.body; }
    var el = window.document.createElement('iframe');
    var css = 'visibility:hidden;width:0;height:0;position:absolute;z-index:-9999;bottom:0;';
    el.setAttribute('src', 'about:blank');
    el.setAttribute('style', css);
    el.setAttribute('width', '0');
    el.setAttribute('height', '0');
    el.setAttribute('wmode', 'opaque');
    parent.appendChild(el);
    return el;
}
var Printd = /** @class */ (function () {
    function Printd(parent) {
        if (parent === void 0) { parent = window.document.body; }
        this.node = null;
        this.parent = parent;
        this.iframe = createIFrame(this.parent);
    }
    Printd.prototype.getIFrame = function () {
        return this.iframe;
    };
    Printd.prototype.print = function (el, cssText, callback) {
        this.node = el.cloneNode(true);
        var _a = this.iframe, contentDocument = _a.contentDocument, contentWindow = _a.contentWindow;
        if (cssText) {
            contentDocument.head.appendChild(createStyle(contentDocument, cssText));
        }
        if (this.node) {
            contentDocument.body.innerHTML = '';
            contentDocument.body.appendChild(this.node);
            if (callback) {
                callback(contentWindow, contentDocument, this.node, this.launchPrint);
            }
            else {
                this.launchPrint(contentWindow);
            }
        }
    };
    Printd.prototype.launchPrint = function (contentWindow) {
        var result = contentWindow.document.execCommand('print', false, null);
        if (!result) {
            contentWindow.print();
        }
    };
    return Printd;
}());

exports.Printd = Printd;
exports.createIFrame = createIFrame;
exports.createStyle = createStyle;
exports['default'] = Printd;

Object.defineProperty(exports, '__esModule', { value: true });

})));
